storage/store_pool.go

```diff
@@ -0,0 +1,298 @@
+// and descriptor.
+func (sd *storeDetail) markAlive(foundAliveOn time.Time, storeDesc proto.StoreDescriptor) {
+	sd.desc = storeDesc
+	sd.dead = false
+	sd.lastUpdatedTime = foundAliveOn
+}
+
+// storePoolPQ implements the heap.Interface (which includes  sort.Interface)
+// and holds storeDetail.
+type storePoolPQ []*storeDetail
+
+// Len implements the sort.Interface.
+func (pq storePoolPQ) Len() int {
+	return len(pq)
+}
+
+// Less implements the sort.Interface.
+func (pq storePoolPQ) Less(i, j int) bool {
+	return pq[i].lastUpdatedTime.UnixNano() < pq[j].lastUpdatedTime.UnixNano()
+}
+
+// Swap implements the sort.Interface.
+func (pq storePoolPQ) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index, pq[j].index = i, j
+}
+
+// Push implements the heap.Interface.
+func (pq *storePoolPQ) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*storeDetail)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+// Pop implements the heap.Interface.
+func (pq *storePoolPQ) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// peek returns the next value in the priority queue without dequeuing it.
+// Lock is assumed held by caller.
+func (pq *storePoolPQ) peek() *storeDetail {
+	if len(*pq) == 0 {
+		return nil
+	}
+	detail := (*pq)[0]
+	return detail
+}
+
+// enqueue either adds the detail to the queue or updates its location in the
+// priority queue.
+// Lock is assumed held by caller.
+func (pq *storePoolPQ) enqueue(detail *storeDetail) {
+	if detail.index < 0 {
+		heap.Push(pq, detail)
+	} else {
+		heap.Fix(pq, detail.index)
+	}
+}
+
+// dequeue removes the next detail from the priority queue.
+// Lock is assumed help by the caller.
+func (pq *storePoolPQ) dequeue() *storeDetail {
+	if len(*pq) == 0 {
+		return nil
+	}
+	detail := heap.Pop(pq).(*storeDetail)
+	return detail
+}
+
+// StorePool maintains a list of all known stores in the cluster and
+// information on their health.
+type StorePool struct {
+	gossip             *gossip.Gossip
+	clock              *hlc.Clock
+	timeUntilStoreDead time.Duration
+
+	mu     sync.RWMutex // Protects variables below...
+	stores map[proto.StoreID]*storeDetail
+	queue  storePoolPQ
+}
+
+// NewStorePool creates a StorePool and registers the store updating callback
+// with gossip.
+func NewStorePool(g *gossip.Gossip, timeUntilStoreDead time.Duration, stopper *stop.Stopper) *StorePool {
+	sp := &StorePool{
+		timeUntilStoreDead: timeUntilStoreDead,
+		stores:             make(map[proto.StoreID]*storeDetail),
+	}
+	heap.Init(&sp.queue)
+
+	storeRegex := gossip.MakePrefixPattern(gossip.KeyStorePrefix)
+	g.RegisterCallback(storeRegex, sp.storeGossipUpdate)
+
+	sp.start(stopper)
+
+	return sp
+}
+
+// storeGossipUpdate The gossip callback used to keep the StorePool up to date.
+func (sp *StorePool) storeGossipUpdate(_ string, content []byte) {
+	storeDesc := new(proto.StoreDescriptor)
```

Comment:
you're making an unnecessary copy here on L184. why not `var storeDesc proto.StoreDescriptor` here and pass a reference into `Unmarshal`? You'll save one copy of this structure on the heap.